torch/csrc/distributed/c10d/c10d_frontend.cpp

```diff
@@ -0,0 +1,635 @@
+int64_t DistributedC10d::getGlobalRank(
+    const std::shared_ptr<ProcessGroup>& group,
+    const int64_t global_rank) const {
+  TORCH_CHECK(
+      group != default_pg_,
+      "group.WORLD does not have local rank to global rank mapping");
+
+  auto it = pg_group_ranks_.find(group);
+  if (it == pg_group_ranks_.end()) {
+    throw std::runtime_error("The given group does not exist");
+  }
+
+  auto& group_rank_map = it->second;
+  for (auto p : group_rank_map) {
+    if (p.second == global_rank) {
+      return p.first;
+    }
+  }
+
+  throw std::runtime_error("The group rank is not part of the group");
+}
+
+std::string DistributedC10d::getBackend(
+    const std::shared_ptr<ProcessGroup>& group) {
+  if (rankNotInGroup(group)) {
+    throw std::runtime_error("Invalid process group specified");
+  }
+
+  auto it = pg_map_.find(group);
+  if (it == pg_map_.end()) {
+    throw std::runtime_error("The given group does not exist");
+  }
+
+  return it->second.first;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::isend(
+    at::Tensor tensor,
+    int64_t dst,
+    const std::shared_ptr<ProcessGroup>& group,
+    c10::optional<int64_t>& tag) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  std::vector<at::Tensor> inputs = {std::move(tensor)};
+
+  checkDefaultPg();
+  if (group == default_pg_) {
+    return default_pg_->send(inputs, dst, tag.value_or(0));
+  }
+
+  auto group_dst_rank = getGroupRank(group, dst);
+  return group->send(inputs, group_dst_rank, tag.value_or(0));
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::irecv(
+    at::Tensor tensor,
+    int64_t src,
+    const std::shared_ptr<ProcessGroup>& group,
+    c10::optional<int64_t>& tag) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  std::vector<at::Tensor> inputs = {std::move(tensor)};
+
+  checkDefaultPg();
+  if (group == default_pg_) {
+    return default_pg_->recv(inputs, src, tag.value_or(0));
+  }
+
+  auto group_dst_rank = getGroupRank(group, src);
+  return group->recv(inputs, group_dst_rank, tag.value_or(0));
+}
+
+void DistributedC10d::send(
+    at::Tensor tensor,
+    int64_t dst,
+    const std::shared_ptr<ProcessGroup>& group,
+    c10::optional<int64_t>& tag) {
+  auto work = isend(std::move(tensor), dst, group, tag);
+  if (work) {
+    work->wait();
+  }
+}
+
+int64_t DistributedC10d::recv(
+    at::Tensor tensor,
+    const c10::optional<int64_t>& src,
+    const std::shared_ptr<ProcessGroup>& group,
+    c10::optional<int64_t>& tag) {
+  if (rankNotInGroup(group)) {
+    return -1;
+  }
+
+  std::vector<at::Tensor> outputs = {std::move(tensor)};
+  if (!src.has_value()) {
+    auto work = group->recvAnysource(outputs, tag.value_or(0));
+    work->wait();
+    auto src_rank = work->sourceRank();
+    if (group == default_pg_) {
+      return src_rank;
+    }
+
+    return getGlobalRank(group, src_rank);
+  }
+
+  if (group == default_pg_) {
+    group->recv(outputs, src.value(), tag.value_or(0))->wait();
```

Comment:
src is an optional so you need to guard it before accessing `value()`?

Reply:
Its value presence is checked online 185.